keps/sig-node/5018-dra-adminaccess/README.md

```diff
@@ -0,0 +1,805 @@
+
+```go
+func (s resourceClaimTemplateStrategy) Validate(ctx context.Context, obj runtime.Object) field.ErrorList {
+	resourceClaimTemplate := obj.(*resource.ResourceClaimTemplate)
+	allErrs := authorizedForAdmin(ctx, resourceClaimTemplate, s.namespaceRest)
+	return append(allErrs, validation.ValidateResourceClaimTemplate(resourceClaimTemplate)...)
+}
+```
+
+and add `authorizedForAdmin` to check if the request is authorized to get admin access to devices based on the DRA admin namespace label:
+
+```go
+func authorizedForAdmin(ctx context.Context, template *resource.ResourceClaimTemplate, namespaceRest NamespaceGetter) field.ErrorList {
+	// omitted
+
+	for i := range template.Spec.Spec.Devices.Requests {
+		value := template.Spec.Spec.Devices.Requests[i].AdminAccess
+		if value != nil && *value {
+			adminRequested = true
+			break
+		}
+	}
+	if !adminRequested {
+		// No need to validate unless admin access is requested
+		return allErrs
+	}
+	if namespaceRest == nil {
+		return append(allErrs, field.Forbidden(field.NewPath(""), "namespace store is nil"))
+	}
+
+	namespaceName := template.Namespace
+	// Retrieve the namespace object from the store
+	obj, err := namespaceRest.Get(ctx, namespaceName, &metav1.GetOptions{ResourceVersion: "0"})
+	if err != nil {
+		return append(allErrs, field.Forbidden(field.NewPath(""), "namespace object cannot be retrieved"))
+	}
+	ns, ok := obj.(*core.Namespace)
+	if !ok {
+		return append(allErrs, field.Forbidden(field.NewPath(""), "namespace object is not of type core.Namespace"))
+	}
+	if value, exists := ns.Labels[DRAAdminNamespaceLabel]; !(exists && value == "true") {
+		return append(allErrs, field.Forbidden(field.NewPath(""), "admin access to devices is not allowed in namespace without DRA Admin Access label"))
+	}
+
+	return allErrs
+}
+```
+
+### Kube-controller-manager Changes
+
+In pkg/controller/resourceclaim/controller.go, process `ResourceClaim` in `syncClaim` function to check for the `adminAccess` field and bypass allocation checks if `adminAccess: true`.
+
+```go
+func (ec *Controller) syncClaim(ctx context.Context, namespace, name string) error {
+	logger := klog.LoggerWithValues(klog.FromContext(ctx), "claim", klog.KRef(namespace, name))
+	ctx = klog.NewContext(ctx, logger)
+	claim, err := ec.claimLister.ResourceClaims(namespace).Get(name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.V(5).Info("nothing to do for claim, it is gone")
+			return nil
+		}
+		return err
+	}
+	adminRequested := false
+	// Check for adminAccess flag
+	for i := range claim.Spec.Devices.Requests {
+		value := claim.Spec.Devices.Requests[i].AdminAccess
+		if value != nil && *value {
+			adminRequested = true
+			break
+		}
+	}
+	if adminRequested {
+		logger.V(5).Info("ResourceClaim", klog.KRef(claim.Namespace, claim.Name), "has admin access, bypass standard allocation checks")
```

Comment:
this is probably not the place for a code review 😜, but I'd want to see this at v=2